Extend VarCouldBeVal Rule to Include Class/Object Properties

sowmyav24 opened the pull request:
Fixes [#1330 ](https://github.com/arturbosch/detekt/issues/1330)

- VarCouldBeVal would run for the class/object variables also
- Excluded variables having lateinit modifiers as they could be injected. Lateinit modifers are verified in `LateinitUsage` rule already.
- Removed specific check for `function.isSomehowNested` . The corresponding tests still pass. Please let me know if this is needed for some other reason

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +112,83 @@ class VarCouldBeValSpec : Spek({
+               }
+            }
+            """
+            assertThat(subject.lint(code)).hasSize(0)
```
nit: use `isEmpty()` instead of `hasSize(0)` for all occurrences.

sowmyav24 replied:
Replaced it the tests

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +112,83 @@ class VarCouldBeValSpec : Spek({
+               }
+            }
+            """
+            assertThat(subject.lint(code)).hasSize(1)
```
You could use the `compileAndLint()` instead of the `lint()` function that additionally checks whether the given test snippet is valid Kotlin code.

sowmyav24 replied:
Done .

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +112,83 @@ class VarCouldBeValSpec : Spek({
+            val code = """
+            class Test {
+               private val someVar = 2
+               private var viewModel = BaseViewModel()
```
Using standard types (e.g. `Int, String, List...`) let's us use the `compileAndLint()` function. Please see my next comment.

sowmyav24 replied:
Sure, Made the changes

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +112,83 @@ class VarCouldBeValSpec : Spek({
+        it("reports variables that are not re-assigned at class level but same variable name is present in a function") {
+            val code = """
+            class Test {
+               private var viewModel = 3
```
Using standard variable names instead of very specific names like `viewModel` helps us to keep the test snippets as simple (to understand) as possible.

sowmyav24 replied:
Done !

sowmyav24 commented:
> Anyhow, I think we should merge this, but add some more test cases for this rule. I'm thinking of nested/inner classes and shadowed variables. What do you think @3flex ?

Shadowed variable test was covered. Renamed it properly. Added test for inner class as well

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -118,26 +119,35 @@ class VarCouldBeValSpec : Spek({
         }
 
-        it("reports variables that are not re-assigned at class level but same variable name is present in a function") {
+        it("reports variables that are not re-assigned at class level but is shadowed in a function") {
```
are shadowed

sowmyav24 replied:
Done !

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -149,7 +159,28 @@ class VarCouldBeValSpec : Spek({
+            }
+        }
+
+        it("reports variables that are not but only used in inner class") {
```
that are only used in inner class

sowmyav24 replied:
Done !

amitdash291 commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +113,113 @@ class VarCouldBeValSpec : Spek({
+            assertThat(subject.compileAndLint(code)).isEmpty()
+        }
+
+        it("does not report variables that are re-assigned in some member function1") {
```
Is this a duplicate of the test right above?

amitdash291 commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:
```diff
@@ -112,6 +113,113 @@ class VarCouldBeValSpec : Spek({
+            val result = subject.compileAndLint(code)
+            assertThat(result).hasSize(1)
+            with(result[0].entity) {
+                assertThat(result[0].entity.ktElement?.text).isEqualTo("private var someVar = 3")
```
Here `result[0].entity.ktElement?.text` can just simply be written as `ktElement?.text`. 
Reason being within the scope of a `with` function,  `result[0].entity` need not be called explicitly. It can be referred to as `this` or even implicitly.

sowmyav24 replied:
Yes, Changed it 👍

sowmyav24 commented:
Before we merge this PR, I wanted to run this rule over the detekt codebase itself. 
```
Run detekt on itself and other kotlin projects with the --run-rule RuleSet:RuleId option to test your rule in isolation
```
is mentioned in `CONTRIBUTING.md`. But i am unable to get it working. Is there any pre requisite to be done before this step ?

sowmyav24 commented:
> Before we merge this PR, I wanted to run this rule over the detekt codebase itself.
> 
> ```
> Run detekt on itself and other kotlin projects with the --run-rule RuleSet:RuleId option to test your rule in isolation
> ```
> 
> is mentioned in `CONTRIBUTING.md`. But i am unable to get it working. Is there any pre requisite to be done before this step ?

Got in running :) .

sowmyav24 commented:
> I would merge this into 1.2.0 after the 1.1.1 hotfix release.
I tried running this rule over detekt codebase. There were a bunch of issues being reported, Where few of them seem to be false negative.
Example: 
- All parameter variables in `CliArgs.kt`
- Parameters in `Detekt.kt` that was set() methods defined like `input` and other similar issues.

schalkms commented:
@sowmyav24 do you have any updates on this PR? Have you tried using the script `scripts/get_analysis_projects.groovy`, which tests rules against 3rd party codebases?

sowmyav24 commented:
@schalkms  I'll try to look into this this weel

imanushin commented:
Probably there are several more cases to exclude:
1. ```override``` property
2. Property with custom attribute, which will be used via reflection (I suggest this option in configuration).

However probably these items can be omitted to the next PRs
